src/client/entity-editor/common/new-date-field.js

```diff
@@ -0,0 +1,154 @@
+
+
+class DateField extends React.Component {
+	constructor(props) {
+		super(props);
+
+		const {day, month, year} = this.props.defaultValue;
+		this.state = {
+			day: !day ? '' : day,
+			month: !month ? '': month,
+			year: !year ? '' : year,
+			warn: dateIsBefore(getTodayDate(),{day, month, year})
+		};
+	}
+
+	updateDate = (day, month, year) => {
+		this.props.onChangeDate({
+			year: !year ? '' : year,
+			month: !month ? '' : month,
+			day: !day ? '' : day
+		});
+
+		this.setState({warn: dateIsBefore(getTodayDate(),{day, month, year})});
+	};
+
+	handleYearChange = (event) => {
+		let year = event.target.value;
+		this.setState({year});
+		this.updateDate(this.state.day, this.state.month, year);
+	};
+
+	handleMonthChange = (event) => {
+		let month = event.target.value;
+		this.setState({month});
+		this.updateDate(this.state.day, month, this.state.year);
+	};
+
+	handleDayChange = (event) => {
+		let day = event.target.value;
+		this.setState({day})
+		this.updateDate(day, this.state.month, this.state.year);
+	};
+
+	handleChangeOfDatePicker = (value) => {
+		const date = new Date(value);
+		const year = (date.getFullYear()).toString();
+		const month = (date.getMonth()+1).toString();
+		const day = (date.getDate()).toString();
+		this.setState({year});
+		this.setState({month});
+		this.setState({day})
+		this.updateDate(day, month, year);
+	}
+
+	render() {
+		
+		const labelElement = (
+			<ValidationLabel
+				empty={this.props.empty}
+				error={this.props.error}
+				warn={this.state.warn}
+				errorMessage={this.props.errorMessage}
+			>
+				{this.props.label}
+			</ValidationLabel>
+		);
+
+		const selectedDate = dateObjectToString({
+			year: this.state.year,
+			month: this.state.month,
+			day: this.state.day
+		});
+		const momentDate = moment(selectedDate);
+		const groupClassName = classNames({hidden: !this.props.show});
+		return (
+			<div>
+				<CustomInput
+					groupClassName={groupClassName}
+					label={labelElement}
+				>
+					<InputGroup style={{width: '17em'}}>
+						<FormControl
+							maxLength="4"
+							placeholder="YYYY"
+							type="text"
+							value={this.state.year}
+							onChange={this.handleYearChange}
+						/>
+					<InputGroup.Addon style={{padding: "0 0.5em"}}>-</InputGroup.Addon>
+						<FormControl
+							maxLength="2"
+							placeholder="MM"
+							style={{width: '3.5em'}}
+							type="text"
+							value={this.state.month}
+							onChange={this.handleMonthChange}
+						/>
+					<InputGroup.Addon style={{padding: "0 0.5em"}}>-</InputGroup.Addon>
+						<FormControl
+							maxLength="2"
+							placeholder="DD"
+							style={{width: '3.5em'}}
+							type="text"
+							value={this.state.day}
+							onChange={this.handleDayChange}
+						/>
+					<InputGroup.Button style={{ fontSize: "inherit"}}>
+							<DatePicker
+								peekNextMonth
+								showMonthDropdown
+								showYearDropdown
+								selected={momentDate.isValid() ? momentDate : null}
```

Comment:
You can replace moment entirely here and use a small utility function to detect if a date is valid:
```
function isValidDate(d) {
  return d instanceof Date && !isNaN(d);
}
```

(Not mine, found it on [StackOverflow](https://stackoverflow.com/a/1353711))